pkg/drivers/kvm/network.go

```diff
@@ -67,52 +73,175 @@ func setupNetwork(conn *libvirt.Connect, name string) error {
+
+	// network: private
+
+	// Only create the private network if it does not already exist
+	if _, err := conn.LookupNetworkByName(d.PrivateNetwork); err != nil {
+		// create the XML for the private network from our networkTmpl
+		tmpl := template.Must(template.New("network").Parse(networkTmpl))
+		var networkXML bytes.Buffer
+		if err := tmpl.Execute(&networkXML, d); err != nil {
+			return errors.Wrap(err, "executing network template")
+		}
+
+		// define the network using our template
+		network, err := conn.NetworkDefineXML(networkXML.String())
 		if err != nil {
-			return errors.Wrap(err, "generating mac address")
+			return errors.Wrapf(err, "defining network from xml: %s", networkXML.String())
+		}
+
+		// and finally create it
+		if err := network.Create(); err != nil {
+			return errors.Wrapf(err, "creating network %s", d.PrivateNetwork)
 		}
-		d.MAC = mac.String()
 	}
+
+	return nil
+}
+
+func (d *Driver) deleteNetwork() error {
+	type source struct {
+		//XMLName xml.Name `xml:"source"`
+		Network string `xml:"network,attr"`
+	}
+	type iface struct {
+		//XMLName xml.Name `xml:"interface"`
+		Source source `xml:"source"`
+	}
+	type result struct {
+		//XMLName xml.Name `xml:"domain"`
+		Name       string  `xml:"name"`
+		Interfaces []iface `xml:"devices>interface"`
+	}
+
 	conn, err := getConnection()
 	if err != nil {
 		return errors.Wrap(err, "getting libvirt connection")
 	}
 	defer conn.Close()
 
-	tmpl := template.Must(template.New("network").Parse(networkTmpl))
-	var networkXML bytes.Buffer
-	if err := tmpl.Execute(&networkXML, d); err != nil {
-		return errors.Wrap(err, "executing network template")
-	}
+	// network: default
+	// It is assumed that the OS manages this network
 
-	// Start the default network
-	log.Infof("Setting up network %s", defaultNetworkName)
-	if err := setupNetwork(conn, defaultNetworkName); err != nil {
-		return err
+	// network: private
+	log.Debugf("Checking if network %s exists...", d.PrivateNetwork)
+	network, err := conn.LookupNetworkByName(d.PrivateNetwork)
+	if err != nil {
+		// TODO: decide if we really wanna throw an error?
+		return errors.Wrap(err, "network %s does not exist")
 	}
+	log.Debugf("Network %s exists", d.PrivateNetwork)
 
-	//Check if network already exists
-	if _, err := conn.LookupNetworkByName(d.PrivateNetwork); err == nil {
-		return nil
+	// iterate over every (also turned off) domains, and check if it
+	// is using the private network. Do *not* delete the network if
+	// that is the case
+	log.Debug("Trying to list all domains...")
+	doms, err := conn.ListAllDomains(0)
+	if err != nil {
+		return errors.Wrap(err, "list all domains")
 	}
+	log.Debugf("Listed all domains: total of %d domains", len(doms))
 
-	network, err := conn.NetworkDefineXML(networkXML.String())
-	if err != nil {
-		return errors.Wrapf(err, "defining network from xml: %s", networkXML.String())
+	// fail if there are 0 domains
+	if len(doms) == 0 {
+		return fmt.Errorf("list of domains is 0 lenght")
```

Comment:
nit: typo. What's the correct fix here? It would be good to include some steps the user could take to create the domain.

Reply:
yeah ... good point to think about what this error means to the user ... it's one of those statements that will/should never happen ... I think there is two things we could do to address this:
1. turn this into a `panic()`
2. simply remove the check: at the end of the day, it would just mean that something else deleted even our own minikube domain while we're trying to delete this network, and then there is no point of further iteration through the domains.

I'd personally opt for (2), but I let you make that decision @dlorenc. What do you think?

Reply:
Let's not panic: https://github.com/golang/go/wiki/CodeReviewComments#dont-panic

I like your #2 option, since this is just at the deletion stage, but just in case it helps a future debugger, do you mind putting a log.Warningf in about the unexpected case?

Thanks again for your contribution.